activerecord/lib/active_record/validations/uniqueness.rb

```diff
@@ -65,26 +65,24 @@ def build_relation(klass, table, attribute, value) #:nodoc:
           attribute_name = attribute.to_s
         end
 
-        column = klass.columns_hash[attribute_name]
         cast_type = klass.type_for_attribute(attribute_name)
         value = cast_type.serialize(value)
         value = klass.connection.type_cast(value)
+
+        if value.nil?
+          return klass.unscoped.where(table[attribute].eq(Arel::Nodes::Quoted.new(value)))
+        end
+
+        column = klass.columns_hash[attribute_name]
         if value.is_a?(String) && column.limit
           value = value.to_s[0, column.limit]
         end
 
-        comparison = if !options[:case_sensitive] && !value.nil?
-          # will use SQL LOWER function before comparison, unless it detects a case insensitive collation
-          klass.connection.case_insensitive_comparison(table, attribute, column, value)
-        else
-          klass.connection.case_sensitive_comparison(table, attribute, column, value)
-        end
-        if value.nil?
-          klass.unscoped.where(comparison)
-        else
-          bind = Relation::QueryAttribute.new(attribute.to_s, value, Type::Value.new)
-          klass.unscoped.where(comparison, bind)
-        end
+        comparison = klass.connection.uniqueness_comparison(
```

Comment:
Construct the BindParam here, and then pass it as `value`. That way, it gets created in a place where its real value is known, and the relationship between the two is clear.